From 94c8ff4a7d268ef8ed7cb854501d9277d72455fa Mon Sep 17 00:00:00 2001
From: arathinx <ayyappanx.rathinam@intel.com>
Date: Wed, 9 Oct 2019 17:14:27 +0800
Subject: [PATCH] PONRTSYS-4429: If SyncE enabled, SFP link to Paragon-X is
 broken after reboot

CGU driver needs to reset the DPLL_CFG5 register to default value 0
---
 drivers/clk/intel/clk-cgu-pll.c | 6 ++++++
 drivers/clk/intel/clk-prx300.c  | 7 +++++++
 2 files changed, 13 insertions(+)

diff --git a/drivers/clk/intel/clk-cgu-pll.c b/drivers/clk/intel/clk-cgu-pll.c
index 06e98961da43..11503bfe1374 100644
--- a/drivers/clk/intel/clk-cgu-pll.c
+++ b/drivers/clk/intel/clk-cgu-pll.c
@@ -454,6 +454,12 @@ void intel_clk_plls_parse_vco_config(struct intel_clk_provider *ctx,
 			break;
 
 		clk_id   = r.args[0];
+		/* All clk rates in table are divided by 1000 of the
+		 * real clk rate to support clk rate upto 10G, So expecting
+		 * the same from dts to match with clk rate table.
+		 * Actual clk rate setting only will use the mult, div, frac
+		 * value from clk rate table.
+		 */
 		clk_rate = r.args[1];
 		clk_en   = r.args[2];
 
diff --git a/drivers/clk/intel/clk-prx300.c b/drivers/clk/intel/clk-prx300.c
index 5023a628b480..856b71dd2577 100644
--- a/drivers/clk/intel/clk-prx300.c
+++ b/drivers/clk/intel/clk-prx300.c
@@ -30,6 +30,7 @@
 #define CGU_GATE1		0x120
 #define CGU_GATE2		0x130
 #define CGU_IF_CLK		0x140
+#define DPLL_CFG5		0x214
 
 #define PLL_DIV(x)		((x) + 0x4)
 #define PLL_SSC(x)		((x) + 0x10)
@@ -307,6 +308,12 @@ static int __init intel_prx300_cgu_probe(struct platform_device *pdev)
 
 	intel_clk_register_plls(ctx, prx300_pll_clks,
 				ARRAY_SIZE(prx300_pll_clks));
+	/* PON FW enables the sync (DPLL_CFG5 register configs),
+	 * After soft reboot ETH link between SFP and Paragon-X cannot be
+	 * established anymore only power-cycle of the SFP helps.
+	 * CGU driver needs to reset DPLL_CFG5 register and re-init the LJPLL3
+	 */
+	intel_set_clk_val(map, DPLL_CFG5, 0, 31, 0);
 	intel_clk_plls_parse_vco_config(ctx, prx300_pll_clks,
 					ARRAY_SIZE(prx300_pll_clks));
 	intel_clk_register_branches(ctx, prx300_branch_clks,
